plugins/binanceExchange_ws.go

```diff
@@ -0,0 +1,283 @@
+
+func isStale(data mapData, ttl time.Duration) bool {
+
+	return time.Now().Sub(data.createdAt).Seconds() > ttl.Seconds()
+}
+
+type mapEvents struct {
+	data map[string]mapData
+	mtx  *sync.Mutex
+}
+
+func (m *mapEvents) Set(key string, data interface{}) {
+
+	now := time.Now()
+
+	m.mtx.Lock()
+	m.data[key] = mapData{
+		data:      data,
+		createdAt: now,
+	}
+	m.mtx.Unlock()
+}
+
+func (m *mapEvents) Get(key string) (mapData, bool) {
+	m.mtx.Lock()
+	data, isData := m.data[key]
+	m.mtx.Unlock()
+
+	return data, isData
+}
+
+func (m *mapEvents) Del(key string) {
+	m.mtx.Lock()
+	delete(m.data, key)
+	m.mtx.Unlock()
+
+}
+
+func makeMapEvents() *mapEvents {
+	return &mapEvents{
+		data: make(map[string]mapData),
+		mtx:  &sync.Mutex{},
+	}
+}
+
+type events struct {
+	SymbolStats *mapEvents
+}
+
+func createStateEvents() *events {
+	events := &events{
+		SymbolStats: makeMapEvents(),
+	}
+
+	return events
+}
+
+// subscribe for symbol@ticker
+func subcribeTicker(symbol string, state *mapEvents) (*stream, error) {
+
+	wsMarketStatHandler := func(ticker *binance.WsMarketStatEvent) {
+		state.Set(symbol, ticker)
+	}
+
+	errHandler := func(err error) {
+		log.Printf("Error WsMarketsStat for symbol %s: %v\n", symbol, err)
+	}
+
+	doneC, stopC, err := binance.WsMarketStatServe(symbol, wsMarketStatHandler, errHandler)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &stream{doneC: doneC, stopC: stopC, cleanup: func() {
+		state.Del(symbol)
+	}}, err
+
+}
+
+type binanceExchangeWs struct {
+	events    *events
+	delimiter string
+
+	streams map[string]*stream
+
+	assetConverter model.AssetConverterInterface
+
+	mtx *sync.Mutex
+}
+
+// makeBinanceWs is a factory method to make an binance exchange over ws
+func makeBinanceWs() (*binanceExchangeWs, error) {
+
+	binance.WebsocketKeepalive = true
+
+	events := createStateEvents()
+
+	beWs := &binanceExchangeWs{
+		events:         events,
+		delimiter:      "",
+		assetConverter: model.CcxtAssetConverter,
+		mtx:            &sync.Mutex{},
+		streams:        make(map[string]*stream),
+	}
+
+	return beWs, nil
+}
+
+func getPrecision(floatStr string) int8 {
+
+	strs := strings.Split(floatStr, ".")
+
+	if len(strs) != 2 {
+		log.Printf("error get precision for float %s\n", floatStr)
+		return 0
+	}
+
+	return int8(len(strs[1]))
+}
+
+// GetTickerPrice impl.
+func (beWs *binanceExchangeWs) GetTickerPrice(pairs []model.TradingPair) (map[model.TradingPair]api.Ticker, error) {
+
+	priceResult := map[model.TradingPair]api.Ticker{}
+	for _, p := range pairs {
+
+		symbol, err := p.ToString(beWs.assetConverter, beWs.delimiter)
+
+		if err != nil {
+			return nil, err
+		}
+
+		tickerData, isTicker := beWs.events.SymbolStats.Get(symbol)
+
+		if !isTicker {
+			stream, err := subcribeTicker(symbol, beWs.events.SymbolStats)
+
+			if err != nil {
+				return nil, fmt.Errorf("error when subscribing for %s: %s", symbol, err)
+			}
+
+			//Store stream
+			beWs.mtx.Lock()
+			beWs.streams[symbol+STREAMTICKER] = stream
```

Comment:
for all streams, it would be better if we did this:
```golang
STREAM_TICKER_FMT = "%s@ticker"
beWs.streams[fmt.Sprintf(STREAM_TICKER_FMT, symbol)] = stream
```
that way it's very clear in the STREAM_TICKER_FMT variable where we will be inserting the variable. This also makes it a little easier to reference in the binance API.